File: 2019/Homework 7/161044116/hw7/src/hw7/BoardArray2D.java
package hw7;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Scanner;

public class BoardArray2D extends AbstractBoard {
	
	String[][] puzzle2;
	int[][] puzzlesize = new int[100][100];
	String[][] correctpuzzle2;
	int totalmoves2D = 0;	
	char lastmovee= 'S';
	
	public BoardArray2D()
	{
		numberofboards++;
	}
	/**
	 * create the right board.
	 */
	void reset_correct() 
	{
		int sayac = 0;
		ArrayList<String> temppuzzle2 = new ArrayList<>();
		for (int i = 0; i < column; ++i)
		{
			for (int k = 0; k < row; ++k)
			{
				if(!puzzle2[i][k].equals("00") && !puzzle2[i][k].equals("bb") )
				{
					temppuzzle2.add(new String( puzzle2[i][k] ));
					sayac++;
				}
			}	
		}
		Collections.sort(temppuzzle2);
		sayac = 0;
		correctpuzzle2 = new String[column][row];
		for (int i = 0; i < column; ++i)
		{
			for (int k = 0; k < row; ++k)
			{
				if (puzzle2[i][k].equals("00") )
				{
					correctpuzzle2[i][k]= "00";
				}
				else if (i == (column-1) && (k ==row-1))
					correctpuzzle2[column-1][row-1]= "bb";
				else
				{
					correctpuzzle2[i][k] = temppuzzle2.get(sayac);
					sayac++;
				}
			}	
		}
	}	
	/**
	 * Resets the board to the solution.
	 */
	@Override
	void reset() 
	{
		reset_correct();
		for (int i = 0; i < column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				puzzle2[i][j] =correctpuzzle2[i][j];
			}
		}	
	}
	/**
	 * @param movee sends the desired move
	 */
	@Override
	void move(char movee) {
		lastmovee = movee;
		String temp;
		switch(movee)
		{
			case 'R':
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j].equals("bb"))
							{
								if(!puzzle2[i][j+1].equals("00")  && j!=row-1)
								{
									
									puzzle2[i][j] = puzzle2[i][j+1];
									puzzle2[i][j+1] = "bb";
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									System.out.println("this move (R) can not be made");
							}
						}
					}	
			break;
			case 'L':
					
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j].equals("bb"))
							{
								if(!puzzle2[i][j-1].equals("00")  && j!=0)
								{
					                temp = puzzle2[i][j-1];
					                puzzle2[i][j-1] = puzzle2[i][j];
					                puzzle2[i][j] = temp;
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									System.out.println("this move (L) can not be made");
							}
						}
					}
			break;
			case 'U':
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j].equals("bb"))
							{
								if(!puzzle2[i-1][j].equals("00")  && i!=0)
								{
									temp = puzzle2[i-1][j];
									puzzle2[i-1][j] = puzzle2[i][j];
									puzzle2[i][j] = temp;
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									System.out.println("this move (U) can not be made");
							}
						}
					}
			break;
			case 'D':
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j].equals("bb"))
							{
								if(!puzzle2[i+1][j].equals("00")  && i!=column-1)
								{
									temp = puzzle2[i][j];
									puzzle2[i][j] = puzzle2[i+1][j];
									puzzle2[i+1][j] = temp;
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									System.out.println("this move (D) can not be made");
							}
						}
					}
			break;
			default:
				System.out.println("Invalid Choice...");
		}
	}
	/**
	 * @return return  board is true or false 
	 */
	@Override
	boolean isSolved() {
		boolean finish = true;
		for (int i = 0; i < column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				if(!puzzle2[i][j].equals(correctpuzzle2[i][j]))
				{
					return false;
				}
			}
		}
		return finish;
	}
	/**
	 * @return data in given indexs
	 * @param i index i
	 * @param y index y
	 */
	@Override
	String cell(int i, int y) {
		
		return puzzle2[i][y];
	}
	/**
	 * @return return String 
	 */
	@Override
	public String toString() 
	{
		String str ="";
		for(int i = 0; i<column ;i++)
		{
			for(int k = 0; k<row ;k++) 
			{
				str += puzzle2[i][k];
				str+=" ";
			}
			str += "\n";
		}	
		return str;
	}
	/**
	 * @return last move  
	 */
	@Override
	char lastMove()
	{	
		return lastmovee;
	}
	/**
	 * @return number of moves  
	 */
	@Override
	int numberOfMoves()
	{
		return totalmoves2D;
	}
	/**
	 * @param filename write  
	 * @exception FileNotFoundException when there is no file
	 */
	@Override
	void writeToFile(String filename) throws IOException{
		 BufferedWriter writer = new BufferedWriter(new FileWriter(filename));
		    writer.write(toString());
		    writer.close();	
	}
	/**
	 * @param filename reading  
	 * @exception FileNotFoundException when there is no file
	 */
	@Override
	void readFromFile(String filename) throws FileNotFoundException 
	{
		File text = new File("./src/hw7/"+filename);
		try (Scanner scnr = new Scanner(text)) {
			int lineNumber = 0;
			int satir = 0;
			String[] temp = new String[1000];
			while(scnr.hasNextLine())
			{
			    String line = scnr.nextLine();
			    String[] ff = line.split(" ");             
			    
			    for(int i = 0;i < 9;i++)
			    {
			    	temp[lineNumber] = ff[i];
			    	lineNumber++;
			    }
			    satir++;
			}
			int sutun =lineNumber/satir;
			int sayac = 0;
			
			puzzle2 = new String[satir][sutun];
			correctpuzzle2 = new String[satir][sutun];
			
			for(int k =0 ;k <satir; k++) 
			{
				for(int j =0 ;j <sutun; j++) 
				{
					puzzle2[k][j] = temp[sayac];
					sayac++;
				}
				set_row_column(sutun,satir);
			
			}
		}
	}
	/**
	 * @param other the compared board
	 * @return Boards are equals or not equals 
	 */
	boolean Equals(BoardArray2D other)
	{
		String[][] board1 = this.puzzle2;
		String[][] board2 = other.puzzle2;
		if(!Arrays.equals(board1,board2))
			return false;
		return true;
	}

}

File: 2019/Homework 5/161044116/boardarray2d.cpp
#include <iostream>
#include <time.h>
#include <cstdlib>
#include <string>
#include <fstream>
#include <memory>

#include "boardarray2d.h"

using std::cout;
using std::cin;

namespace NPuzzle
{	
	BoardArray2D:: BoardArray2D(): AbstractBoard(){
		numberofboardss++;
		set_row_column( 3,3);
		reset_variable();
		cout<<endl<<numberofboardss<<". 2D BOARD CREATED!!!"<<endl<<endl;
	}
	string BoardArray2D::operator ()(int x,int y)
	{
		return puzzle2[x][y];
	}
	BoardArray2D::~BoardArray2D()
	{
		for(int i = 0; i < column; ++i){

			  delete [] puzzle2[i];
		}
		delete []puzzle2;

		for(int i = 0; i < column; ++i){

			  delete [] correctpuzzle2[i];
		}
		delete []correctpuzzle2;
	}
	void BoardArray2D::reset_variable()
	{
		puzzle2 = new string*[column];
		correctpuzzle2 = new string*[column];
		for (int i = 0; i < column; ++i)
		{
			puzzle2[i] = new string[row];
			correctpuzzle2[i] = new string[row];
		}
	}
	void BoardArray2D::setSize(int x ,int y)
	{
		puzzlesize = new int*[x*y];
		cout<<"WORKING SETSIZE  2D FUNCTION.."<<endl;
		for (int i = 0; i < x; ++i)
		{
			puzzlesize[i] = new int[y];
		}	
		int key =1;
		for (int i = 0; i < x; ++i)
		{
			for (int j = 0; j < y; ++j)
			{
				if(i==(x-1) && j ==(y-1))
					puzzlesize[x-1][y-1] = 0;
				else
				{
					puzzlesize[i][j] = key;
					cout<<puzzlesize[i][j]<<"  ";		
				}
				if(key<10)
					cout<<" ";
				key++;			
			}
			cout<<endl;
		}
	}
	void BoardArray2D::readFromFile(string takefile)
	{

		filename = takefile;
	 	int i,t;
	 	string rowto;
	 	string  temppuzzle[1000];	
		file.open(filename.c_str());
		if(file.is_open())
		{
			while(getline(file,line))
			{
				c_temp++;
			}								
			file.clear();
			file.seekg( 0, std::ios::beg);			
		
			while(!file.eof())
			{
				file >> rowto;
				total++;
			}
			total--;

			r_temp = total/c_temp;
			set_row_column(r_temp,c_temp);
			reset_variable();
			t=0;
			file.clear();
			file.seekg( 0, std::ios::beg);			

			while(file>>str)
			{
				temppuzzle[t] = str;
				t++;
			}




			t = 0;
			string key;
			
			for (int i = 0; i < column; ++i)
			{
				for (int j = 0; j < row; ++j)
				{
					key = temppuzzle[t];
					puzzle2[i][j] = key;
					t++;
				}
			}
		}	
		cout <<endl;
		return;
	}
	void BoardArray2D::writeToFile(string takefile)
	{
		savefilename = takefile;
		savef.open(savefilename.c_str());
		int temp =0;
		for (int i = 0; i <column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				savef << puzzle2[i][j];
				savef << " ";
				temp++;
			}
			savef << endl;
		}
		savef.close();
		cout<<endl<<"The game has been saved."<<endl;
	}
	void BoardArray2D::print()
	{
		cout<<"---------------------------"<<endl;	
		for (int j = 0; j < column ; ++j) 
		{	
			for (int i = 0; i < row ; ++i)
			{
				cout <<puzzle2[j][i]<<" ";
			}
			cout<<endl;
		}
	}
	void BoardArray2D::reset_correct()
	{
		int sayac = 0;
		
		string * temppuzzle2;
		temppuzzle2 = new string[row*column];
		
		for (int i = 0; i < column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				if(puzzle2[i][j]!= "00" && puzzle2[i][j]!= "bb")
				{
					temppuzzle2[sayac] = puzzle2[i][j];
					sayac++;
				}
			}
		}
		for (int i = 0; i < sayac; ++i)
		{
			int j;
			string key;
			key = temppuzzle2[i];
			j = i-1;
			while(j>=0 && temppuzzle2[j] > key)
			{
				temppuzzle2[j+1] = temppuzzle2[j];
				j=j-1;
			}
			temppuzzle2[j+1]=key;
		}

		sayac = 0;
		for (int i = 0; i < column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				if (puzzle2[i][j] == "00")
				{
					correctpuzzle2[i][j] = puzzle2[i][j];
				}
				else
				{
					correctpuzzle2[i][j] = temppuzzle2[sayac];
					sayac++;
				}
			}
			
		}
		correctpuzzle2[column-1][row-1]= "bb";
	}
	void BoardArray2D::reset()
	{
		reset_correct();
		for (int i = 0; i < column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				puzzle2[i][j] =correctpuzzle2[i][j];
			}
		}
	}
	bool BoardArray2D::isSolved()
	{
		int finish;
		
		for (int i = 0; i < column; ++i)
		{
			for (int j = 0; j < row; ++j)
			{
				if(puzzle2[i][j] == correctpuzzle2[i][j])
				{
					finish = 0;
				}
				else
				{ 
					finish = -1;
					i = total;
				}
			}
			
		}
		if (finish == 0)
			return true;
		else
			return false;
	}
	void BoardArray2D::move(char movee)
	{
		lastmovee = movee;
		string temp;
		switch(movee)
		{
			case 'R':
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j]=="bb")
							{
								if(puzzle2[i][j+1]!= "00"  && j!=row-1)
								{
									puzzle2[i][j] = puzzle2[i][j+1];
									puzzle2[i][j+1] = "bb";
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									cout<<"this move (R) can not be made"<<endl;
							}
						}
					}
			break;
			case 'L':
					
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j]=="bb")
							{
								if(puzzle2[i][j-1]!= "00"  && j!=0)
								{
									temp;
					                temp = puzzle2[i][j-1];
					                puzzle2[i][j-1] = puzzle2[i][j];
					                puzzle2[i][j] = temp;
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									cout<<"this move (L) can not be made"<<endl;
							}
						}
					}
			break;
			case 'U':
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j]=="bb")
							{
								if(puzzle2[i-1][j]!= "00"  && i!=0)
								{
									cout<<"ceee";
									temp = puzzle2[i-1][j];
									puzzle2[i-1][j] = puzzle2[i][j];
									puzzle2[i][j] = temp;
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									cout<<"this move (U) can not be made"<<endl;
							}
						}
					}
			break;
			case 'D':
					for (int i = 0; i < column; ++i)
					{
						for(int j= 0; j < row ; ++j)
						{
							if(puzzle2[i][j]=="bb")
							{
								if(puzzle2[i+1][j] != "00"  && i!=column-1)
								{
									temp = puzzle2[i][j];
									puzzle2[i][j] = puzzle2[i+1][j];
									puzzle2[i+1][j] = temp;
									i=column;
									j=row;
									totalmoves2D++;
								}
								else
									cout<<"this move (D) can not be made"<<endl;
							}
						}
					}
			break;
			default:
					cout<<"INVALID CHOICE.."<<endl;
		}
	}
	char BoardArray2D::lastMove()
	{
		return lastmovee;
	}
	int BoardArray2D::numberOfMoves()
	{
		return totalmoves2D;
	}
}